File: 05_arrays/05_21_three_sum/ThreeSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeSum {
    public static void main(String[] args) {
        int[] arr = {-1, 0, 1, 2, -1, -4};
        System.out.println(threeSum(arr));
    }

    private static List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < nums.length - 1; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int low = i + 1;
            int high = nums.length - 1;
            while (low < high) {
                int sum = nums[i] + nums[low] + nums[high];
                if (sum < 0) {
                    low++;
                } else if (sum > 0) {
                    high--;
                } else {
                    result.add(List.of(nums[i], nums[low], nums[high]));
                    while (low + 1 < high && nums[low] == nums[low + 1]) {
                        low++;
                    }
                    while (high - 1 > low && nums[high] == nums[high - 1]) {
                        high--;
                    }
                    low++;
                    high--;
                }
            }
        }
        return result;
    }
}

File: 05_arrays/05_21_three_sum/three_sum.py
def three_sum(nums: list):
    result = []
    nums.sort()
    
    for i in range(len(nums)):
        if i > 0 and nums[i] == nums[i-1]:
            continue
        low = i + 1
        high = len(nums) - 1
        
        while low < high:
            sumation = nums[i] + nums[low] + nums[high]
            if sumation < 0:
                low += 1
            elif sumation > 0: 
                high -= 1
            else:
                result.append([nums[i], nums[low], nums[high]])
                while low + 1 < high and nums[low] == nums[low + 1]:
                    low += 1
                while high - 1 > low and nums[high] == nums[high - 1]:
                    high -= 1
                low += 1
                high -= 1
    
    return result


print(three_sum([-1, 0, 1, 2, -1, -4]))